sdks/node-sdk/src/utils/streams.ts

```diff
@@ -0,0 +1,199 @@
+   * The delay between retries (in milliseconds)
+   * (default: 10000)
+   */
+  retryDelay?: number;
+  /**
+   * Whether to retry the stream if it fails
+   * (default: true)
+   */
+  retryOnFail?: boolean;
+};
+
+export type StreamCallback<T = unknown> = (
+  error: Error | null,
+  value: T | undefined,
+) => void;
+
+export type StreamFunction<T = unknown> = (
+  callback: StreamCallback<T>,
+  onFail: () => void,
+) => Promise<StreamCloser>;
+
+export type StreamValueMutator<T = unknown, V = T> = (
+  value: T,
+) => V | Promise<V>;
+
+/**
+ * Creates a stream from a stream function
+ *
+ * If the stream fails, an attempt will be made to restart it.
+ *
+ * This function is not intended to be used directly.
+ *
+ * @param streamFunction - The stream function to create a stream from
+ * @param streamValueMutator - An optional function to mutate the value emitted from the stream
+ * @param options - The options for the stream
+ * @param args - Additional arguments to pass to the stream function
+ * @returns An async iterable stream proxy
+ * @throws {StreamInvalidRetryAttemptsError} if the retryAttempts option is less than 0 and retryOnFail is true
+ * @throws {StreamFailedError} if the stream fails and can't be restarted
+ */
+export const createStream = async <T = unknown, V = T>(
+  streamFunction: StreamFunction<T>,
+  streamValueMutator?: StreamValueMutator<T, V>,
+  options?: StreamOptions<T, V>,
+) => {
+  const {
+    onError,
+    onFail,
+    onRestart,
+    onRetry,
+    onValue,
+    retryAttempts = DEFAULT_RETRY_ATTEMPTS,
+    retryDelay = DEFAULT_RETRY_DELAY,
+    retryOnFail = true,
+  } = options ?? {};
+  // retry attempts must be greater than 0
+  if (retryOnFail && retryAttempts < 0) {
+    throw new StreamInvalidRetryAttemptsError();
+  }
+
+  const asyncStream = new AsyncStream<V>();
+  const streamCallback: StreamCallback<T> = (error, value) => {
+    // if a stream error occurs, call the onError callback
+    if (error) {
+      onError?.(error);
+      return;
+    }
+    // ensure the value is not undefined
+    if (value !== undefined) {
+      try {
+        // if a streamValueMutator is provided, mutate the value
+        if (streamValueMutator) {
+          const mutatedValue = streamValueMutator(value);
+          if (isPromise(mutatedValue)) {
+            void mutatedValue
+              .then((mutatedValue) => {
+                asyncStream.push(mutatedValue);
+                onValue?.(mutatedValue);
+              })
+              .catch((error: unknown) => {
+                onError?.(error as Error);
+              });
+          } else {
+            asyncStream.push(mutatedValue);
+            onValue?.(mutatedValue);
+          }
+        } else {
+          asyncStream.push(value as unknown as V);
+          onValue?.(value as unknown as V);
+        }
+      } catch (error) {
+        onError?.(error as Error);
+      }
+    }
+  };
+  const retry = async (retries: number = retryAttempts) => {
```

Comment:
does this have the issue where if a retry happens while there is no connection, the stream starts without detecting an issue even though it is not connected? Or is it attempting to do a sync somewhere upstream, I think you might have mentioned, for determining if we're online to get around that issue? 

I ask because I was seeing that problem when implementing retry for streams in react native, if I retried while connection was down, it thought it was working and it would never flag the stream ending.